keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,882 @@
+  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
+-->
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
+Look at the [existing SLIs/SLOs].
+
+Think about adding additional work or introducing new steps in between
+(e.g. need to do X to start a container), etc. Please describe the details.
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Share Resources with Service API
```

Comment:
To complete this though (because unifying is so compelling absent the gory details), I'll sketch some more thoughts.  I'm not 100%, but I think I am leaning against this.

Suppose we define `CIDR` as an API type:

```
type CIDR struct {
    Purpose CIDRPurpose // Services, Pods, ...
}
```

The name of the CIDR is the CIDR itself.  E.g.  `10.0.0.0/16`.

To allocate a sub-range, you create a `CIDRClaim` object, such as `10.0.1.0/24`.

The ServiceAllocator finds all CIDRs where `purpose` == `Service`.  It always allocates a /32 or /128 address.

The NodePodIPAMAllocator would need an indirection to specify which nodes can use which ranges and masks.  E.g.

```
type PodCIDRAllocationPolicy struct {
    NodeSelector  // these nodes
    CIDRSelector  // can use these CIDRs
    Mask int     // with this mask size
}
```

That's a fair bit of complexity, and it trades 2 resources (ServiceCIDR, PodCIDR) for 2 resources (PodCIDR, PodCIDRAllocationPolicy).  It doesn't really make anything easier - the controllers are actually MORE complicated now.  And it doesn't actually solve the NodePort allocation problem, so we're stuck with another resource, but different in form and function.

It's not clear to me whether there's any other use case for IP or CIDR allocation or for port allocation that helps to amortize this cost.  If there are other use-cases, the balance could shift.

Ergo, my lean towards "generalizing is not worth the effort".

Reply:
:upside_down_face: